skeleton/protractor/aurelia_project/tasks/protractor.ext

```diff
@@ -15,13 +26,17 @@ gulp.task('protractor', (cb) => {
   // @if feat.typescript
   gulp.src('test/e2e/**/*.e2e.ts')
   // @endif
-    .pipe(protractor({configFile: 'test/protractor.conf.js'}))
+    .pipe(protractor({configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
     .on('error', cb)
-    .on('end', cb);
+    .on('end', () => {
+      cb()
+      process.exit(0);
```

Comment:
Same issue as the above cypress exit. The exit code needs to respect failing, especially to support CI automation.

Reply:
In this case, it could also change to the following to respect the failing tests: 

```ts
gulp.task('protractor', (cb) => {
  gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(
      protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
    .on('error', () =>{
      cb('Failed Tests');
      console.log("Finished with errors");
      process.exit(-1);
    })
    .on('end', () => {
      cb();
      console.log("Finished Successfully");
      process.exit(0);
    });
});
```

However, as far as I tested, I might be wrong though correct me please, the **on_end** will not get called if we have failed tests, instead, it calls the **on_error**. but I totally agree to make a clear distinction with `process.exit` to respect the passed and failed tests.

Reply:
The same scenarios goes here, I tested this one in a `Dockerfile` and `protractor` itself returns a non-zero code when it faces failed tests, so the following will suffice, IMO: 

```ts
gulp.task('protractor', (cb) => {
  gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(
      protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
    .on('error', cb)
    .on('end', () => {
      cb();
      console.log("Finished Successfully");
      process.exit(0);
    });
});
```

Reply:
`process.exit(-1);` it should be a positive number for exit code, at least for POSIX OS.
http://www.tldp.org/LDP/abs/html/exitcodes.html

Reply:
The previous `on('error', cb)` will just work as expected, because it passes error to cb callback.

But the whole protractor is outdated using gulp v3 way. In gulp v4, we need to simplify it to:
```js
import gulp from 'gulp';
const {protractor, webdriver_update }  = require('gulp-protractor');

function runProtractor() {
  return gulp.src('test/e2e/**/*.e2e.js')
    .pipe(protractor({configFile: 'test/protractor.conf.js'}));
}

// Setting up the test task
export default gulp.series(
  webdriver_update,
  runProtractor
);
```
I tested it, pass or fail as expected.

Reply:
Does it automatically finish the whole process when running in headless mode and all tests passed? On my machine it just waits for the user when all tests pass!

```
au protrctor --start --run --port 4040
```

**protractor.ts**
```ts
import * as gulp from 'gulp';
const { protractor, webdriver_update } = require('gulp-protractor');

import { CLIOptions } from 'aurelia-cli';
import { default as run } from './run';

function runApp(cb) {
  if (CLIOptions.hasFlag('start')) {
    run();
  }
  cb();
}

function runProtractor() {
  return gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
  ;
}

export default gulp.series(
  runApp,
  webdriver_update,
  runProtractor
);
```


![image](https://user-images.githubusercontent.com/4405518/61167432-2e0dc100-a554-11e9-8e0b-9c4e2c52e50f.png)

Reply:
You new protractor task is waiting for the `run` dev server. That's why it didn't stop.

We talked about this on cypress too. We were looking into native API on web-dev-server and browsersync to close the dev server properly.

The "Process exited with error code 1" is the message from a sub-process of protractor testing (remember you cannot get args from `au protractor` inside protractor config js), **not the process of `au protractor`** where you forced exit code 0. You can check the final exit code of `au protractor` with command `echo $?` (or something equivalent on windows).

Reply:
> We talked about this on cypress too. We were looking into native API on web-dev-server and browsersync to close the dev server properly.

I reverted all changes ( in the discussion, not pushed up to the repo yet ) on the failure case as it stopped and exited everything automatically, the failure case is exactly like the first day, both on `cypress` and `protractor`. 

For the success path I thought having `process.exit(0);` has no harm, right?, and it is way simpler than hacking around with passing the `server` between scripts, but if you think in the success path it will be better that we ask the server to terminate rather than exiting the process with exit code 0, it will change to the following:

**protractor.ts**

```ts
import * as gulp from 'gulp';
const { protractor, webdriver_update } = require('gulp-protractor');

import { CLIOptions } from 'aurelia-cli';
import { runWebpack as runWebpack } from './run';

let _server;

function runApp(cb) {
  if (CLIOptions.hasFlag('start')) {
    runWebpack((server) => {
      _server = server;
    });
  }
  cb();
}

function runProtractor(cb) {
  return gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
    .on('error', cb)
    .on('end', () => {
      cb();
      _server.close();
    });
  ;
}

// Setting up the test task
export default gulp.series(
  runApp,
  webdriver_update,
  runProtractor
);

```

**run.ts**
```ts
import * as webpack from 'webpack';
import * as Server from 'webpack-dev-server';
import * as project from '../aurelia.json';
import * as gulp from 'gulp';

import { config } from './build';
import configureEnvironment from './environment';
import { CLIOptions, reportWebpackReadiness } from 'aurelia-cli';

function _runWebpack(done) {
  // https://webpack.github.io/docs/webpack-dev-server.html
  let opts = {
    host: CLIOptions.getFlagValue('host') || project.platform.host,
    port: CLIOptions.getFlagValue('port') || project.platform.port,
    open: CLIOptions.hasFlag('open') || project.platform.open,
    publicPath: config.output.publicPath,
    filename: config.output.filename,
    hot: project.platform.hmr || CLIOptions.hasFlag('hmr'),
    contentBase: config.output.path,
    historyApiFallback: true,
    stats: {
      colors: require('supports-color')
    },
    ...config.devServer
  } as any;

  // Add the webpack-dev-server client to the webpack entry point
  // The path for the client to use such as `webpack-dev-server/client?http://${opts.host}:${opts.port}/` is not required
  // The path used is derived from window.location in the browser and output.publicPath in the webpack.config.
  if (project.platform.hmr || CLIOptions.hasFlag('hmr')) {
    config.plugins.push(new webpack.HotModuleReplacementPlugin());
    config.entry.app.unshift('webpack-dev-server/client', 'webpack/hot/dev-server');
  } else {
    // removed "<script src="/webpack-dev-server.js"></script>" from index.ejs in favour of this method
    config.entry.app.unshift('webpack-dev-server/client');
  }

  const compiler = webpack(config);
  let server = new Server(compiler, opts);

  server.listen(opts.port, opts.host, function (err) {
    if (err) throw err;

    reportWebpackReadiness(opts);
    done(server);
  });
}

function runWebpack(done) {
  return _runWebpack(() => done());
}

const run = gulp.series(
  configureEnvironment,
  runWebpack
);

export { run as default, _runWebpack as runWebpack };

```

Reply:
This should work. Please test it :-)

BTW, I am pretty sure you can simplify
```js
function runProtractor(cb) {
  return gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
    .on('error', cb)
    .on('end', () => {
      cb();
      _server.close();
    });
  ;
}
```

To
```js
function runProtractor() {
  return gulp.src('test/e2e/**/*.e2e.ts')
    .pipe(protractor({ configFile: 'protractor.conf.js', args: process.argv.slice(3) }))
    .on('end', () => {
      _server.close(); // or _server.exit() for browserSync
    });
}
```

Gulp knows how to it if return value is a gulp stream.

Reply:
The server close should help to finally exit the `au protractor` command without the need of `process.exit`. At least I hope so.